### DISCUSSION TITLE
opencv4nodejs

### DISCUSSION THREAD

**Discussion by cherishman2005** [cherishman2005/rtc#44] [General]
Ahh yeah, my fault sorry. From opencv version 3.4.2 upwards we don't export some functions anymore such as the one your mentioned above:

#if CV_MINOR_VERSION < 2
  Nan::SetPrototypeMethod(ctor, "addTrainingSample", AddTrainingSample);
  Nan::SetPrototypeMethod(ctor, "addTrainingSampleAsync",
	  AddTrainingSampleAsync);
  Nan::SetPrototypeMethod(ctor, "getData", GetData);
  Nan::SetPrototypeMethod(ctor, "getDataAsync", GetDataAsync);
  Nan::SetPrototypeMethod(ctor, "getFaces", GetFaces);
  Nan::SetPrototypeMethod(ctor, "getFacesAsync", GetFacesAsync);
  Nan::SetPrototypeMethod(ctor, "setFaceDetector", SetFaceDetector);
  Nan::SetPrototypeMethod(ctor, "training", Training);
  Nan::SetPrototypeMethod(ctor, "trainingAsync", TrainingAsync);
#endif
Since they changed the facemark API with 3.4.2 the package wouldn't compile anymore, that's why I removed them. If you want to use the aboveshown functions, you can still use them with 3.4.1, or submit a PR with bindings to the new base class.

TLDR: Probably the example simply has to be adjusted and the calls to setFaceDetector and getFaces have to be removed.

**Comment by cherishman2005** [comment]
Installing a Specific Version of OpenCV
You can specify the Version of OpenCV you want to install via the script by setting an environment variable: export OPENCV4NODEJS_AUTOBUILD_OPENCV_VERSION=4.1.0

**Comment by cherishman2005** [comment]
root@ubuntu:/home/nodejs/face_detection# export OPENCV4NODEJS_AUTOBUILD_OPENCV_VERSION=3.4.1
root@ubuntu:/home/nodejs/face_detection# npm install

> opencv-build@0.1.9 install /home/nodejs/face_detection/node_modules/opencv-build
> node ./install.js

info install if you want to use an own OpenCV installation set OPENCV4NODEJS_DISABLE_AUTOBUILD
info readAutoBuildFile file does not exists: /home/nodejs/face_detection/node_modules/opencv-build/opencv/auto-build.json /home/nodejs/face_detection/node_modules/opencv-build/opencv/auto-build.json
info install failed to find auto-build.json: /home/nodejs/face_detection/node_modules/opencv-build/opencv/auto-build.json
info install 
info install running install script...
info install 
info install opencv version: 3.4.1
info install with opencv contrib: yes
info install custom build flags: 
info install 
info install executing: git --version
info install git --version: git version 2.7.4
info install 
info install executing: cmake --version
info install cmake --version: cmake version 3.5.1
info install 
info install CMake suite maintained and supported by Kitware (kitware.com/cmake).
info install 
info install installing opencv version 3.4.1 into directory: /home/nodejs/face_detection/node_modules/opencv-build/opencv
Cloning into 'opencv_contrib'...
remote: Enumerating objects: 2142, done.
remote: Counting objects: 100% (2142/2142), done.
remote: Compressing objects: 100% (1843/1843), done.
remote: Total 2142 (delta 371), reused 1241 (delta 190), pack-reused 0
Receiving objects: 100% (2142/2142), 50.69 MiB | 1.55 MiB/s, done.
Resolving deltas: 100% (371/371), done.
Checking connectivity... done.
Note: checking out 'ced5aa760688dd2ec867ebf7bd4f0c2341d2fde5'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

Cloning into 'opencv'...
remote: Enumerating objects: 6901, done.
remote: Counting objects: 100% (6901/6901), done.
remote: Compressing objects: 100% (5890/5890), done.
remote: Total 6901 (delta 1076), reused 4221 (delta 684), pack-reused 0
Receiving objects: 100% (6901/6901), 80.18 MiB | 1.62 MiB/s, done.
Resolving deltas: 100% (1076/1076), done.
Checking connectivity... done.
Note: checking out '6ffc48769ac60d53c4bd1913eac15117c9b1c9f7'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

**Comment by cherishman2005** [comment]
```
export OPENCV4NODEJS_AUTOBUILD_OPENCV_VERSION=3.4.1
"opencv4nodejs": "^4.7.2"
```

facemark.js
```
const cv = require("opencv4nodejs");
const fs = require("fs");
const path = require("path");

if (!cv.xmodules.face) {
  throw new Error("exiting: opencv4nodejs compiled without face module");
}

const facemarkModelPath = "./data/face/";
const modelFile = path.resolve(facemarkModelPath, "lbfmodel.yaml");

if (!fs.existsSync(modelFile)) {
  console.log("could not find landmarks model");
  console.log(
    "download the model from: https://raw.githubusercontent.com/kurnianggoro/GSOC2017/master/data/lbfmodel.yaml"
  );
  throw new Error("exiting: could not find landmarks model");
}

const classifier = new cv.CascadeClassifier(cv.HAAR_FRONTALFACE_ALT2);

// create the facemark object with the landmarks model
const facemark = new cv.FacemarkLBF();
facemark.loadModel(modelFile);

const image = cv.imread("./data/got.jpg");
const gray = image.bgrToGray();

var faceClassifierOpts = {
    minSize: new cv.Size(30, 30),
    scaleFactor: 1.126,
    minNeighbors: 1,
}

const faces = classifier.detectMultiScale(gray, faceClassifierOpts).objects
// use the detected faces to detect the landmarks
const faceLandmarks = facemark.fit(gray, faces);

for (let i = 0; i < faceLandmarks.length; i++) {
  const landmarks = faceLandmarks[i];
  for (let x = 0; x < landmarks.length; x++) {
    image.drawCircle(landmarks[x], 1, new cv.Vec(0, 255, 0), 1, cv.LINE_8);
  }
}

//cv.imshowWait("VideoCapture", image);
cv.imwrite("got_out.jpg", image);
```

输入：
![got](https://user-images.githubusercontent.com/17688273/149066128-6b0ab55b-42d8-4f43-bb47-4546e317fbae.jpg)


输出：
![got_out](https://user-images.githubusercontent.com/17688273/149066144-9f3d6664-c242-4054-a499-519a73e907dc.jpg)
